Add support for python2 and python3 (no minor in version)

beleki opened the pull request:
fix #7

beleki committed:
Support python2, python3 and python3.7

- new entrypoints run_in_virtualenv2, run_in_virtualenv3, run_in_virtualenv3.7
- add matrix to travis
- add check-manifest<0.42, devpi-client incompability

vereszol commented on src/virtualenvrunner/pythonversionrun.py:
```diff
@@ -20,16 +20,22 @@ def _get_run_lambda(self, python_version):
-            minor=python_version.minor,
-            exe_suffix=get_exe_suffix()),
-                  python_version=python_version)
+        if python_version.minor:
```
Instead of `if .. else` it would be better to use a helper function, same way like `get_exe_suffix`.

Even better solution would be to treat version as one entity and implement its string representation in PythonVersion class, something like below:

```
class PythonVersion:
    def __init__(self, major, minor):
        self.major = major
        self.minor = minor

    def __str__(self):
        if not self.minor:
            return str(self.major)
        return '{}.{}'.format(self.major, self.minor)

def get_python_versions():
    for v in [(2, None), (2, 7), (3, None), (3, 4), (3, 5), (3, 6)]:
        yield PythonVersion(*v)

for v in get_python_versions():
    print('python{version}'.format(version=v))
```

beleki replied:
Refactored

vereszol commented on src/virtualenvrunner/pythonversionrun.py:
```diff
@@ -20,16 +20,22 @@ def _get_run_lambda(self, python_version):
             return self._run_functions[name]
         except KeyError:
-            raise AttributeError('{cls} object has no attribute {name}'.format(
+            raise AttributeError('{cls} object has no attribute {name}'.format(  # pylint: disable=raise-missing-from
```
Disabling pylint checks `raise-missing-from` and `super-with-arguments` should be on its own commit, as it is not related to content of this change.

beleki replied:
Added as new commit

vereszol replied:
OK

beleki committed:
Disable some pylint and pep8 checks

- pylint is run with py3 but code must be python2 and python3 compatible
- fix version of pytest-console-scripts==0.1.6, the newer versions support only python 3
- pytest pep8 is removed

vereszol commented on src/virtualenvrunner/python_versions.py:
```diff
@@ -15,4 +15,3 @@ def __str__(self):
 def get_python_versions():
     for v in [(2, None), (2, 7), (3, None), (3, 4), (3, 5), (3, 6)]:
         yield PythonVersion(*v)
-
```
Trailing newline should not be removed

vereszol commented on setup.py:
```diff
@@ -50,9 +50,9 @@ def get_console_scripts():
+        [('{clif.cli}{version} = '
           'virtualenvrunner.cli:{clif.function}{major}{minor}'.format(
-              major=v.major, minor=v.minor, clif=clif))
+             version=str(v), major=v.major, minor=v.minor, clif=clif))
```
No need to use `str()` as `format` is calling that automatically unless some format is specified.

vereszol commented on src/virtualenvrunner/pythonversionrun.py:
```diff
@@ -20,11 +20,10 @@ def _get_run_lambda(self, python_version):
-            major=python_version.major,
-            minor=python_version.minor,
+        self._run(pythonexe='python{version}{exe_suffix}'.format(
+            version=str(python_version),
```
No need to use `str`

vereszol commented on tox.ini:
```diff
@@ -1,15 +1,15 @@
 # Copyright (C) 2019, Nokia
 
 [tox]
-envlist = py27, py36, docs, pylint
```
Should not remove `py36` as it seems to be still supported by the tool and we even provide `run_in_virtualenv3.6`.
On the other hand, adding `py37` is fine, we could even add it to `get_python_versions`.

vereszol replied:
OK

vereszol commented on tox.ini:
```diff
@@ -1,15 +1,15 @@
     pytest-cov
     pytest-pep8
-    pytest-console-scripts
+    pytest-console-scripts==0.1.6
```
Reason for changes to dependencies (and also pytest options) should be explained in commit message.
This (and pytest option changes) could be in same commit where pylint checks are disabled.

vereszol replied:
ok

vereszol commented on tox.ini:
```diff
@@ -31,8 +31,6 @@ commands = {posargs:py.test -v\
            {envsitepackagesdir}/virtualenvrunner {toxinidir}/tests}
 
 [pytest]
-addopts = --pep8
```
If pep8 checks are not used any more you could remove `pytest-pep8` from dependency list.

vereszol replied:
ok

vereszol commented on tox.ini:
```diff
@@ -31,8 +31,6 @@ commands = {posargs:py.test -v\
 
 [pytest]
-addopts = --pep8
-pep8maxlinelength = 90
```
As pep8 is not used any more pylint should be configured to expect same line length at [.pylintrc#L196](https://github.com/nokia/virtualenvrunner/blob/master/.pylintrc#L196)

vereszol replied:
ok
